pkg/controller.v2/controller_service.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2018 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package controller provides a Kubernetes controller for a TFJob resource.
+package controller
+
+import (
+	"fmt"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	tfv1alpha2 "github.com/kubeflow/tf-operator/pkg/apis/tensorflow/v1alpha2"
+)
+
+// reconcileServices checks and updates services for each given TFReplicaSpec.
+// It will requeue the tfjob in case of an error while creating/deleting services.
+func (tc *TFJobController) reconcileServices(
+	tfjob *tfv1alpha2.TFJob,
+	services []*v1.Service,
+	rtype tfv1alpha2.TFReplicaType,
+	spec *tfv1alpha2.TFReplicaSpec) error {
+	tfjobKey, err := KeyFunc(tfjob)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for tfjob object %#v: %v", tfjob, err))
+		return err
+	}
+
+	// Convert TFReplicaType to lower string.
+	rt := strings.ToLower(string(rtype))
+
+	// Get active services for this TFReplicaType.
+	activeServices := filterActiveServicesForTFReplicaType(services, rt)
+
+	diff := len(activeServices) - int(*(spec.Replicas))
+
+	if diff < 0 {
+		// Need to create new services.
+		diffIndexes := getDiffServiceIndexes(activeServices, *spec.Replicas)
+		if diff+len(diffIndexes) != 0 {
+			// This should never happened.
+			return fmt.Errorf("services diff(%d) is not equal to length(%d) of diffIndexes", diff, len(diffIndexes))
+		}
+
+		expectationServicesKey := genExpectationServicesKey(tfjobKey, rt)
+		err := tc.expectations.ExpectCreations(expectationServicesKey, diff)
+		if err != nil {
+			return err
+		}
+
+		for _, index := range diffIndexes {
+			log.Infof("need to create new service: %s-%s", rt, index)
+
+			// Create OwnerReference.
+			controllerRef := genOwnerReference(tfjob)
+
+			// Append tfReplicaTypeLabel and tfReplicaIndexLabel labels.
+			labels := genLabels(tfjobKey)
+			labels[tfReplicaTypeLabel] = rt
+			labels[tfReplicaIndexLabel] = index
+
+			service := &v1.Service{
+				Spec: v1.ServiceSpec{
+					Selector: labels,
+					Ports: []v1.ServicePort{
+						{
+							Name: genGeneralName(tfjobKey, rt, index),
+							Port: 2222,
```

Comment:
I will update it.

Reply:
We removed the field `TFPort`, thus we should reuse the ports in podtemplatespec. And I will open an issue to keep track. Currently we use a hard coded port and do not support customization.

I do not intent to implement in this PR since it is copied from upstream/v1alpha2. I will file a new PR for this feature.

WDYT?

Reply:
https://github.com/kubeflow/tf-operator/issues/532